File: src/providers/search/serper.ts
/**
 * Serper.dev Search Provider
 *
 * Uses Google Search via Serper.dev API.
 * Requires SERPER_API_KEY environment variable.
 */

import type { SearchProvider, SearchResult, SearchOptions } from './types.js';

const API_CONFIG = {
  BASE_URL: 'https://google.serper.dev',
  ENDPOINT: '/search',
  DEFAULT_NUM_RESULTS: 10,
  DEFAULT_TIMEOUT: 10000,
} as const;

interface SerperRequest {
  q: string;
  num?: number;
}

interface SerperOrganicResult {
  title: string;
  link: string;
  snippet: string;
  position: number;
}

interface SerperResponse {
  organic: SerperOrganicResult[];
  searchParameters: {
    q: string;
    type: string;
    num: number;
  };
}

/**
 * Filter results by allowed/blocked domains
 */
function filterByDomain(results: SearchResult[], options?: SearchOptions): SearchResult[] {
  if (!options?.allowedDomains?.length && !options?.blockedDomains?.length) {
    return results;
  }

  return results.filter((result) => {
    try {
      const domain = new URL(result.url).hostname;

      if (options.allowedDomains?.length) {
        return options.allowedDomains.some(
          (allowed) => domain === allowed || domain.endsWith('.' + allowed)
        );
      }

      if (options.blockedDomains?.length) {
        return !options.blockedDomains.some(
          (blocked) => domain === blocked || domain.endsWith('.' + blocked)
        );
      }

      return true;
    } catch {
      return true;
    }
  });
}

export class SerperProvider implements SearchProvider {
  readonly name = 'serper' as const;
  private apiKey: string;

  constructor(apiKey?: string) {
    this.apiKey = apiKey ?? process.env.SERPER_API_KEY ?? '';
    if (!this.apiKey) {
      throw new Error('SERPER_API_KEY environment variable is required for Serper provider');
    }
  }

  async search(query: string, options?: SearchOptions): Promise<SearchResult[]> {
    const searchRequest: SerperRequest = {
      q: query,
      num: options?.numResults ?? API_CONFIG.DEFAULT_NUM_RESULTS,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(
      () => controller.abort(),
      options?.timeout ?? API_CONFIG.DEFAULT_TIMEOUT
    );

    try {
      const signals = options?.abortSignal
        ? [controller.signal, options.abortSignal]
        : [controller.signal];

      const response = await fetch(`${API_CONFIG.BASE_URL}${API_CONFIG.ENDPOINT}`, {
        method: 'POST',
        headers: {
          'X-API-KEY': this.apiKey,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(searchRequest),
        signal: AbortSignal.any(signals),
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Serper search error (${response.status}): ${errorText}`);
      }

      const data = await response.json() as SerperResponse;

      const results: SearchResult[] = (data.organic || []).map((item) => ({
        title: item.title,
        url: item.link,
        snippet: item.snippet,
      }));

      return filterByDomain(results, options);
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof Error && error.name === 'AbortError') {
        throw new Error('Search request timed out');
      }

      throw error;
    }
  }
}

File: internal/provider/search/serper.go
package search

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

const (
	serperEndpoint = "https://google.serper.dev/search"
	serperEnvKey   = "SERPER_API_KEY"
)

// SerperProvider implements the Serper.dev search provider
type SerperProvider struct {
	apiKey string
}

// NewSerperProvider creates a new Serper provider
func NewSerperProvider() *SerperProvider {
	return &SerperProvider{
		apiKey: os.Getenv(serperEnvKey),
	}
}

func (p *SerperProvider) Name() ProviderName    { return ProviderSerper }
func (p *SerperProvider) DisplayName() string   { return "Serper (Google)" }
func (p *SerperProvider) RequiresAPIKey() bool  { return true }
func (p *SerperProvider) EnvVars() []string     { return []string{serperEnvKey} }
func (p *SerperProvider) IsAvailable() bool     { return p.apiKey != "" }

// serperRequest represents a Serper API request
type serperRequest struct {
	Q   string `json:"q"`
	Num int    `json:"num,omitempty"`
}

// serperResponse represents a Serper API response
type serperResponse struct {
	Organic []struct {
		Title   string `json:"title"`
		Link    string `json:"link"`
		Snippet string `json:"snippet"`
	} `json:"organic"`
}

// Search performs a web search using Serper
func (p *SerperProvider) Search(ctx context.Context, query string, opts SearchOptions) ([]SearchResult, error) {
	if !p.IsAvailable() {
		return nil, fmt.Errorf("%s environment variable is not set", serperEnvKey)
	}

	numResults := opts.NumResults
	if numResults <= 0 {
		numResults = 10
	}

	// Build request
	reqBody := serperRequest{
		Q:   query,
		Num: numResults,
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create HTTP request
	client := &http.Client{Timeout: getTimeout(opts)}
	req, err := http.NewRequestWithContext(ctx, "POST", serperEndpoint, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-API-KEY", p.apiKey)

	// Execute request
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(respBody))
	}

	// Read response
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Parse response
	var serperResp serperResponse
	if err := json.Unmarshal(respBody, &serperResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	// Convert to SearchResult and apply domain filtering
	results := make([]SearchResult, 0, len(serperResp.Organic))
	for _, r := range serperResp.Organic {
		// Apply domain filtering (Serper doesn't support it natively)
		if !matchesDomainFilter(r.Link, opts.AllowedDomains, opts.BlockedDomains) {
			continue
		}

		results = append(results, SearchResult{
			Title:   r.Title,
			URL:     r.Link,
			Snippet: truncateSnippet(r.Snippet, 200),
		})
	}

	return results, nil
}
